MultiRaft writes to its input channels after close

The following test appears to have failed:

[#4232](https://circleci.com/gh/cockroachdb/cockroach/4232):

```
I0628 23:55:50.080297     311 range_command.go:774] range 1: new leader lease replica 1:1 00:00:00.000 +1.000s
I0628 23:55:50.080613     311 range.go:1023] gossiping accounting config from store 1, range 1
I0628 23:55:50.081254     311 range.go:1023] gossiping permissions config from store 1, range 1
I0628 23:55:50.081802     311 range.go:1023] gossiping zones config from store 1, range 1
==================
WARNING: DATA RACE
Read by goroutine 80:
  runtime.chansend()
      /usr/src/go/src/runtime/chan.go:83 +0x0
  github.com/cockroachdb/cockroach/storage.(*Store).ProposeRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:1408 +0x23f
  github.com/cockroachdb/cockroach/storage.(*Range).proposeRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:737 +0x5ca
  github.com/cockroachdb/cockroach/storage.(*Range).requestLeaderLease()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:345 +0x55e
  github.com/cockroachdb/cockroach/storage.(*Range).redirectOnOrAcquireLeaderLease()
--
  github.com/cockroachdb/cockroach/storage.func·034()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:575 +0x61
  github.com/cockroachdb/cockroach/util.func·001()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:75 +0x7f

Previous write by goroutine 10:
  runtime.closechan()
      /usr/src/go/src/runtime/chan.go:257 +0x0
  github.com/cockroachdb/cockroach/multiraft.(*state).stop()
      /go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:636 +0x652
  github.com/cockroachdb/cockroach/multiraft.func·004()
      /go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:491 +0x763
  github.com/cockroachdb/cockroach/util.func·001()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:75 +0x7f

Goroutine 80 (running) created at:
--
  github.com/cockroachdb/cockroach/kv.TestTxnCoordSenderGC()
      /go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender_test.go:460 +0xda
  testing.tRunner()
      /usr/src/go/src/testing/testing.go:447 +0x133
==================
panic: send on closed channel

goroutine 1376 [running]:
github.com/cockroachdb/cockroach/storage.(*Store).ProposeRaftCommand(0xc2080da000, 0xc2087d8db0, 0x10, 0x1, 0x100000001, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1408 +0x240
github.com/cockroachdb/cockroach/storage.(*Range).proposeRaftCommand(0xc20875a360, 0x7f148df0b490, 0xc20823cf90, 0x7f148df09708, 0xc2080da300, 0x7f148df0b4d0, 0xc2080116c0, 0x1, 0x630000006c)
	/go/src/github.com/cockroachdb/cockroach/storage/range.go:737 +0x5cb
github.com/cockroachdb/cockroach/storage.(*Range).requestLeaderLease(0xc20875a360, 0x2540be401, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/range.go:345 +0x55f
github.com/cockroachdb/cockroach/storage.(*Range).redirectOnOrAcquireLeaderLease(0xc20875a360, 0x2540be401, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/range.go:385 +0x296
--
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender_test.go:490 +0x89e
testing.tRunner(0xc20875a000, 0x1a2dc98)
	/usr/src/go/src/testing/testing.go:447 +0x134
created by testing.RunTests
	/usr/src/go/src/testing/testing.go:555 +0xd4e
FAIL	github.com/cockroachdb/cockroach/kv	5.419s
=== RUN TestHeartbeatSingleGroup-2
--- PASS: TestHeartbeatSingleGroup-2 (0.41s)
=== RUN TestHeartbeatMultipleGroupsJointLeader-2
--- PASS: TestHeartbeatMultipleGroupsJointLeader-2 (0.05s)
=== RUN TestHeartbeatResponseFanout-2
I0628 23:55:52.448233     322 heartbeat_test.go:425] SubmitCommand succeed after Heartbeat Response fanout
E0628 23:55:52.452318     322 transport.go:202] sending rpc failed: read tcp 127.0.0.1:46854: use of closed network connection
--- PASS: TestHeartbeatResponseFanout-2 (0.08s)
=== RUN TestInitialLeaderElection-2
I0628 23:55:52.455317     322 multiraft_test.go:158] testing leader election for node 0

```
Please assign, take a look and update the issue accordingly.

--- a/multiraft/multiraft.go
+++ b/multiraft/multiraft.go
@@ -36,11 +36,11 @@ const (
 
 // An ErrGroupDeleted is returned for commands which are pending while their
 // group is deleted.
-var ErrGroupDeleted = errors.New("group deleted")
+var ErrGroupDeleted = errors.New("raft group deleted")
 
 // ErrStopped is returned for commands that could not be completed before the
 // node was stopped.
-var ErrStopped = errors.New("stopped")
+var ErrStopped = errors.New("raft processing stopped")
 
 // Config contains the parameters necessary to construct a MultiRaft object.
 type Config struct {
@@ -621,24 +621,6 @@ func (s *state) stop() {
 	}
 	s.MultiRaft.multiNode.Stop()
 	s.MultiRaft.Transport.Stop(s.nodeID)
-
-	// Drain the input channels because other threads may be blocking
-	// on these operations.
-	for {
-		select {
-		case op := <-s.createGroupChan:
-			op.ch <- util.Errorf("shutting down")
-		case op := <-s.removeGroupChan:
-			op.ch <- util.Errorf("shutting down")
-		case op := <-s.proposalChan:
-			op.ch <- util.Errorf("shutting down")
-		default:
-			close(s.createGroupChan)
-			close(s.removeGroupChan)
-			close(s.proposalChan)
-			return
-		}
-	}
 }
 
 // addNode creates a node and registers the given groupIDs for that

--- a/storage/id_alloc.go
+++ b/storage/id_alloc.go
@@ -99,20 +99,19 @@ func (ia *idAllocator) start() {
 					res client.KeyValue
 				)
 				for r := retry.Start(idAllocationRetryOpts); r.Next(); {
-					if ia.stopper.StartTask() {
-						idKey := ia.idKey.Load().(proto.Key)
-						res, err = ia.db.Inc(idKey, int64(ia.blockSize))
-						ia.stopper.FinishTask()
-
-						if err == nil {
-							newValue = res.ValueInt()
-							break
-						}
-
-						log.Warningf("unable to allocate %d ids from %s: %s", ia.blockSize, idKey, err)
-					} else {
+					if !ia.stopper.StartTask() {
 						return
 					}
+					idKey := ia.idKey.Load().(proto.Key)
+					res, err = ia.db.Inc(idKey, int64(ia.blockSize))
+					ia.stopper.FinishTask()
+
+					if err == nil {
+						newValue = res.ValueInt()
+						break
+					}
+
+					log.Warningf("unable to allocate %d ids from %s: %s", ia.blockSize, idKey, err)
 				}
 				if err != nil {
 					panic(fmt.Sprintf("unexpectedly exited id allocation retry loop: %s", err))

--- a/storage/range.go
+++ b/storage/range.go
@@ -943,7 +943,7 @@ func (r *Range) getLeaseForGossip(ctx context.Context) (bool, error) {
 		return false, util.Errorf("no gossip or range not initialized")
 	}
 	if !r.rm.Stopper().StartTask() {
-		return false, util.Errorf("system is shutting down")
+		return false, util.Errorf("node is stopping")
 	}
 	defer r.rm.Stopper().FinishTask()
 
@@ -1053,17 +1053,19 @@ func (r *Range) handleSkippedIntents(args proto.Request, intents []proto.Intent)
 	}
 
 	ctx := r.context()
-	if stopper := r.rm.Stopper(); stopper.StartTask() {
-		go func() {
-			err := r.rm.resolveWriteIntentError(ctx, &proto.WriteIntentError{
-				Intents: intents,
-			}, r, args, proto.CLEANUP_TXN)
-			if wiErr, ok := err.(*proto.WriteIntentError); !ok || wiErr == nil || !wiErr.Resolved {
-				log.Warningc(ctx, "failed to resolve on inconsistent read: %s", err)
-			}
-			stopper.FinishTask()
-		}()
+	stopper := r.rm.Stopper()
+	if !stopper.StartTask() {
+		return
 	}
+	go func() {
+		err := r.rm.resolveWriteIntentError(ctx, &proto.WriteIntentError{
+			Intents: intents,
+		}, r, args, proto.CLEANUP_TXN)
+		if wiErr, ok := err.(*proto.WriteIntentError); !ok || wiErr == nil || !wiErr.Resolved {
+			log.Warningc(ctx, "failed to resolve on inconsistent read: %s", err)
+		}
+		stopper.FinishTask()
+	}()
 }
 
 // TODO(spencerkimball): move to util.

